kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/BigQuerySinkTask.java

```diff
@@ -135,31 +135,75 @@ private void updateAllPartitions(String topic, Map<TopicPartition, OffsetAndMeta
     private final Map<TopicPartition, OffsetAndMetadata> offsets;
     private final String topic;
     private final Set<Schema> schemas;
+    private final BatchWriter<RowToInsert> batchWriter;
 
-    public TableWriter(
-        TableId table,
-        List<RowToInsert> rows,
-        Map<TopicPartition, OffsetAndMetadata> offsets,
-        String topic,
-        Set<Schema> schemas) {
+    public TableWriter(TableId table,
+                       List<RowToInsert> rows,
+                       Map<TopicPartition, OffsetAndMetadata> offsets,
+                       String topic,
+                       Set<Schema> schemas) {
       this.table = table;
       this.rows = rows;
       this.offsets = offsets;
       this.topic = topic;
       this.schemas = schemas;
+      this.batchWriter = batchWriterManager.getBatchWriter(table);
     }
 
     @Override
     public Void call() throws InterruptedException {
-      List<List<RowToInsert>> partitions = rowPartitioner.partition(rows);
-      for (List<RowToInsert> partition : partitions) {
-        bigQueryWriter.writeRows(table, partition, topic, schemas);
-      }
+      batchWriter.writeAll(table, rows, topic, schemas);
       updateAllPartitions(tablesToTopics.get(table), offsets);
       return null;
     }
   }
 
+  private static class BatchWriterManager {
```

Comment:
Ya. I think the parallelism issue could still pop up with multiple partitions, though. If a single task is handling more than one partition for a given table (likely in our setup), there will be a race condition.